internal/cli/csrconfig.go

```diff
@@ -0,0 +1,132 @@
+package cli
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+type CSRConfigOptions struct {
+	GlobalOptions
+	CertSigningRequestFile string
+	Outfile                string
+	Name                   string
+	ExpirationSeconds      string
+}
+
+func DefaultCSRConfigOptions() *CSRConfigOptions {
+	return &CSRConfigOptions{
+		GlobalOptions: DefaultGlobalOptions(),
+	}
+}
+
+func NewCmdCSRConfig() *cobra.Command {
+	o := DefaultCSRConfigOptions()
+	cmd := &cobra.Command{
+		Use:   "csr-generate",
+		Short: "generate a CSR resource config .yaml based on a CSR file .csr",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(cmd, args); err != nil {
+				return err
+			}
+			if err := o.Validate(args); err != nil {
+				return err
+			}
+			return o.Run(cmd.Context(), args)
+		},
+		SilenceUsage: true,
+	}
+	o.Bind(cmd.Flags())
+	return cmd
+}
+
+func (o *CSRConfigOptions) Bind(fs *pflag.FlagSet) {
+	o.GlobalOptions.Bind(fs)
+	fs.StringVarP(&o.Outfile, "outfile", "o", "mygeneratedcsr.yaml", "specify a filename in which to save the generated csr resource config")
+	fs.StringVarP(&o.Name, "name", "n", "mycsr", "specify a name for the csr")
+	fs.StringVarP(&o.ExpirationSeconds, "expiry", "e", "604800", "specify desired certificate expiration in seconds")
+}
+
+func (o *CSRConfigOptions) Complete(cmd *cobra.Command, args []string) error {
+	return o.GlobalOptions.Complete(cmd, args)
+}
+
+func (o *CSRConfigOptions) Validate(args []string) error {
+	if err := o.GlobalOptions.Validate(args); err != nil {
+		return err
+	}
+
+	if len(args) == 0 {
+		return fmt.Errorf("must specify a CSR file")
+	}
+	_, err := os.ReadFile(args[0])
+	if err != nil {
+		return fmt.Errorf("invalid CSR file specified: %w", err)
+	}
+
+	return nil
+}
+
+func (o *CSRConfigOptions) Run(ctx context.Context, args []string) error {
+	if len(args) == 0 {
+		return fmt.Errorf("must specify a CSR file")
+	}
+
+	data, err := os.ReadFile(args[0])
+	if err != nil {
+		return fmt.Errorf("invalid CSR file specified: %w", err)
+	}
+
+	encoded := base64.StdEncoding.EncodeToString(data)
+
+	output1 := `apiVersion: v1alpha1
+kind: CertificateSigningRequest
+metadata:
+  name: `
+	output2 := `
+spec:
+  request: `
+
+	output3 := `
+  signerName: ca
+  usages: ["clientAuth", "CA:false"]
+  expirationSeconds: `
+	filename := createFilename(o.Outfile)
+
+	f, err := os.Create(filename)
```

Comment:
we need to close this file or it will leak a file descriptor.

defer f.Close()

Reply:
good call thanks